transport/internal/tlsmux/conn_sniffer.go

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2022 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tlsmux
+
+import (
+	"bytes"
+	"net"
+)
+
+// connSniffer wraps the connection and enables tlsmux to sniff inital bytes from the
+// connection efficiently.
+type connSniffer struct {
+	net.Conn
+
+	sniffing bool         // true when sniffing.
+	buf      bytes.Buffer // holds sniffed data.
+}
+
+func (c *connSniffer) Read(b []byte) (int, error) {
+	if !c.sniffing && c.buf.Len() != 0 {
+		// Read from the buffer when in non-sniffing mode and buffer is not empty.
+		n, _ := c.buf.Read(b)
+		if c.buf.Len() == 0 {
+			// Release memory as we don't need buffer anymore.
+			c.buf = bytes.Buffer{}
+		}
+		return n, nil
+	}
+
+	n, err := c.Conn.Read(b)
+	if err != nil {
+		return 0, err
+	}
+
+	if c.sniffing {
```

Comment:
Currently we rely on the caller to set sniffing to set true and false explicitly. Can we instead by default set it to true and then toggle it here after storing the buffer?

Reply:
I don't have a firm opinion on this, but we can have a constructor method that sets `sniffing = true`. Though my idea was that the creator should control when sniffing should be enabled or disabled. Generically sniffing can be used more than once (maybe capture TLS renegotiation records).

Reply:
I am not sure if it becomes complicated if we allow multiple sniffing since both the connection might have their own implementation and it could be an inconsistent behaviour in switching the connection unless the bytes are read fully from the connection. Wdyt?

Reply:
At the moment, we don't need in-stream sniffing, so I agree. I have updated the method to sniff by default.